### DISCUSSION TITLE
ot-nrf528xx: ot-cli-{ftd,mtd}: missing joiner command

### DISCUSSION THREAD

**Discussion by stintel** [openthread/openthread#8575] [Q&A] [answered]
Not sure if I'm doing something wrong, or misunderstood the documentation, but I'm trying to join an nRF52840 dongle to a Thread network, but there seem to be a bunch of commands missing.

I built https://github.com/openthread/ot-nrf528xx in a container (openthread/environment), using these steps:

```
mkdir firmware
git clone https://github.com/openthread/ot-nrf528xx.git --recursive
cd ot-nrf528xx
./script/build nrf52840 USB_trans -DOT_BOOTLOADER=USB
arm-none-eabi-objcopy -O ihex build/bin/ot-cli-ftd
arm-none-eabi-objcopy -O ihex build/bin/ot-cli-mtd
wget https://github.com/makerdiary/nrf52840-mdk-usb-dongle/raw/master/tools/uf2conv.py
python3 uf2conv.py ot-cli-ftd   --convert --family 0xADA52840 --output ot-cli-ftd.uf2
python3 uf2conv.py ot-cli-mtd   --convert --family 0xADA52840 --output ot-cli-mtd.uf2
```

Flashed the dongle by double-pressing the button, mounting the filesystem and dropping the uf2 file in there. Then unplug/replug the dongle, and notice the joiner command is missing:

ot-cli-ftd:
```
> joiner
Error 35: InvalidCommand
> help
bbr
bufferinfo
ccathreshold
channel
child
childip
childmax
childtimeout
contextreusedelay
counters
dataset
delaytimermin
detach
discover
dns
domainname
eidcache
eui64
extaddr
extpanid
factoryreset
fem
ifconfig
ipaddr
ipmaddr
joinerport
keysequence
leaderdata
leaderweight
log
mac
mode
multiradio
neighbor
netdata
netstat
networkdiagnostic
networkidtimeout
networkkey
networkname
panid
parent
parentpriority
partitionid
ping
platform
pollperiod
preferrouterid
promiscuous
pskc
rcp
region
releaserouterid
reset
rloc16
router
routerdowngradethreshold
routereligible
routerselectionjitter
routerupgradethreshold
scan
singleton
state
tcp
thread
txpower
udp
unsecureport
version
Done
```

ot-cli-mtd:
```
> joiner
Error 35: InvalidCommand
> help
bbr
bufferinfo
ccathreshold
channel
childtimeout
counters
dataset
detach
discover
dns
domainname
eui64
extaddr
extpanid
factoryreset
fem
ifconfig
ipaddr
ipmaddr
keysequence
leaderdata
log
mac
mode
multiradio
netdata
netstat
networkkey
networkname
panid
parent
ping
platform
pollperiod
promiscuous
rcp
region
reset
rloc16
scan
singleton
state
tcp
thread
txpower
udp
unsecureport
version
Done
```

I can solve this by adding -DOT_JOINER:BOOL=ON when calling the build script, but I wonder if there's a reason for not having this included by default. Hence starting this discussion rather than filing a bug and/or PR.

**Comment by jwhui** [comment, marked as answer]
Thread devices are not required to implement the Joiner role. As a result, the Joiner role is a build feature that is disabled by default.

**Reply by stintel** [reply]
Alright. Just a bit confusing if you're new to the scene I guess.

**Reply by jwhui** [reply]
You may find our External Commissioning guide helpful, which includes a description on [preparing the Joiner device](https://openthread.io/guides/border-router/external-commissioning/prepare#prepare_the_joiner_device).

**Reply by stintel** [reply]
I actually just stumbled on that, as I am having more trouble trying to start the commissioning/joing process via ot-ctl on a host running otbr-agent. Pretty sure this worked a few months ago, but now it doesn't... Was this feature removed and is the use of [openthread/ot-commissioner](https://github.com/openthread/ot-commissioner) now required? I'm pretty sure I was able to do it with just ot-ctl, as I've never built ot-commissioner before.

**Reply by jwhui** [reply]
You can also refer to our [On-Mesh Commissioning guide](https://openthread.io/guides/build/commissioning).

**Reply by stintel** [reply]
That seems to be what I was looking for. Thank you, very much appreciated.
